// @oasisprotocol/sapphire-paratime/lib/cjs/index.cjs
"use strict";
/** @packageDocumentation
 * The main export of this package is {@link wrap}.
 *
 * The {@link cipher} module contains additional ciphers you may use (most notably {@link cipher.Plain}, which can be used for transparency).
 *
 * The {@link signedCalls} module contains utilities for making signed calls that allow the caller to have their address as `msg.sender` during an `eth_call`.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallError = exports.OASIS_CALL_DATA_PUBLIC_KEY = exports.NETWORKS = exports.signedCalls = exports.cipher = void 0;
exports.cipher = __importStar(require("./cipher.cjs"));
__exportStar(require("./compat.cjs"), exports);
exports.signedCalls = __importStar(require("./signed_calls.cjs"));
var networks_js_1 = require("./networks.cjs");
Object.defineProperty(exports, "NETWORKS", { enumerable: true, get: function () { return networks_js_1.NETWORKS; } });
exports.OASIS_CALL_DATA_PUBLIC_KEY = 'oasis_callDataPublicKey';
class CallError extends Error {
    constructor(message, response) {
        super(message);
        this.response = response;
    }
}
exports.CallError = CallError;
